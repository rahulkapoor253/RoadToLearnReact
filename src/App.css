.App {
  text-align: center;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  font-size: calc(8px + 2vmin);
  color: white;
  padding-top: 4vh;
}

.App-link {
  color: #61dafb;
}

.table {
  margin: 20px;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
}
.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}
.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  border: 1px solid #e3e3e3;
}
.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}
.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.button-inline {
  border-width: 0;
  background: rgb(48, 212, 212);
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  font-size: inherit;
  cursor: pointer;
}
.button-active {
  border-radius: 0;
  border-bottom: 1px solid #38bb6c;
}
