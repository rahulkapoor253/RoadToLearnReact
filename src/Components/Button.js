import React from "react";

class Button extends React.Component {
  render() {
    const { children, onClick, className = "" } = this.props;
    return (
      <button type="button" onClick={onClick} className={className}>
        {children}
      </button>
    );
  }
}

export default Button;
