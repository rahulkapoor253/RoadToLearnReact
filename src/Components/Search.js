import React from "react";

//no access to local component state or lifecycle methods, render is also implicit
const Search = ({ value, onChange, children, onSubmit }) => {
  return (
    <form onSubmit={onSubmit}>
      <input
        placeholder={children}
        type="text"
        value={value}
        onChange={onChange}
        maxLength="15"
      />
      <button className="btn-search" type="submit">
        Submit
      </button>
    </form>
  );
};

export default Search;
