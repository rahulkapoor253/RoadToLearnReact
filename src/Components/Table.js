import React from "react";
import Button from "./Button";
import Sort from "./Sort";
import { sortBy } from "lodash";

const SORTS = {
  NONE: (list) => list,
  AUTHOR: (list) => sortBy(list, "author"),
  TITLE: (list) => sortBy(list, "title"),
};

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: "NONE",
    };
  }

  onSort = (sortKey) => {
    this.setState({
      sortKey,
    });
  };

  render() {
    const { list, onDismiss } = this.props;
    const { sortKey } = this.state;
    const sortedList = SORTS[sortKey](list);
    return (
      <div className="table">
        {sortedList.map((item) => (
          <div key={item.objectID} className="table-row">
            <span style={{ width: "70%" }}>
              <a style={{ marginRight: "4px" }} href={item.url}>
                {item.title}
              </a>
              <Sort
                sortKey="TITLE"
                onSort={this.onSort}
                className="button-title"
              />
            </span>
            <span style={{ width: "20%" }}>
              <span style={{ marginRight: "4px" }}>{item.author}</span>
              <Sort sortKey="AUTHOR" onSort={this.onSort} />
            </span>
            <span style={{ width: "10%" }}>
              <Button
                className="button-inline"
                onClick={() => onDismiss(item.objectID)}
                children="Dismiss"
              />
            </span>
          </div>
        ))}
      </div>
    );
  }
}

export default Table;
