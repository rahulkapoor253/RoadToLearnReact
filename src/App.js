import React from "react";
import "./App.css";
import Search from "./Components/Search";
import Table from "./Components/Table";
import Button from "./Components/Button";
import axios from "axios";
//instead of using typescript we can use react proptypes
import PropTypes from "prop-types";
import Loading from "./Components/Loading";
import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  PARAM_HPP,
  PARAM_PAGE,
  PARAM_SEARCH,
  PATH_BASE,
  PATH_SEARCH,
} from "./Constants";

class App extends React.Component {
  _isMounted = false;

  constructor(props) {
    super(props);
    //also store the state of SORT
    this.state = {
      results: null,
      searchKey: "",
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    };

    //auto bound using ES6 arrow functions
    //this.onDismiss = this.onDismiss.bind(this);
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm,
    });

    this.fetchStoriesBySearch(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  fetchStoriesBySearch = (searchTerm, page = 0) => {
    this.setState({
      isLoading: true,
    });

    const proxyurl = "https://cors-anywhere.herokuapp.com/";
    const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`;

    //look for mounting to avoid data leaks
    axios(proxyurl + url)
      .then((response) => {
        if (this._isMounted) {
          this.setTopStoriesResponse(response.data);
          this.setState({
            isLoading: false,
          });
        }
      })
      .catch((error) => {
        if (this._isMounted) {
          this.setState({ error });
          this.setState({
            isLoading: false,
          });
        }
      });
  };

  setTopStoriesResponse = (result) => {
    const { hits, page } = result;
    const { searchKey, results } = this.state;
    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    //console.log(result);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  };

  onSearchSubmit = (event) => {
    event.preventDefault();
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm,
    });
    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchStoriesBySearch();
    }
  };

  onDismiss = (id) => {
    const { results, searchKey } = this.state;
    const { hits, page } = results[searchKey];

    const updatedList = hits.filter((item) => item.objectID !== id);
    this.setState({
      results: { ...results, [searchKey]: { hits: updatedList, page } },
    });
  };

  handleInput = (event) => {
    //console.log(event.target.value);
    this.setState({
      searchTerm: event.target.value,
    });
  };

  render() {
    const { results, searchTerm, searchKey, error } = this.state;
    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;
    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="App">
        <div className="App-header">
          <Search
            value={searchTerm}
            onChange={this.handleInput}
            onSubmit={this.onSearchSubmit}
            children="Search"
          />
          {error ? (
            <div>
              <p>Something went wrong. Please try again.</p>
            </div>
          ) : (
            <React.Fragment>
              {this.state.isLoading ? (
                <div className="loading-container">
                  <Loading />
                  <div>
                    <p>Loading...</p>
                  </div>
                </div>
              ) : (
                <div className="content-container">
                  <Table list={list} onDismiss={this.onDismiss} />
                  <Button
                    children="More"
                    onClick={() =>
                      this.fetchStoriesBySearch(searchKey, page + 1)
                    }
                  />
                </div>
              )}
            </React.Fragment>
          )}
        </div>
      </div>
    );
  }
}

Button.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  onClick: PropTypes.func,
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string,
      author: PropTypes.string,
      url: PropTypes.string,
    })
  ).isRequired,
  onDismiss: PropTypes.func,
};

export default App;
