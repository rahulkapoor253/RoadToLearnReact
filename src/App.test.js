import React from "react";
import renderer from "react-test-renderer";
import ReactDOM from "react-dom";
import App from "./App";
import Search from "./Components/Search";
import Button from "./Components/Button";
import Table from "./Components/Table";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { render, screen } from "@testing-library/react";

//init the adapter
Enzyme.configure({ adapter: new Adapter() });

describe("App", () => {
  it("renders success", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("rendering app component", () => {
    render(<App />);
    //displays skeleton of parent and child components
    //screen.debug();
  });

  test("looking up search in dom", () => {
    render(<App />);
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });
});

describe("Search", () => {
  //it renders search component to dom
  it("renders success", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Search />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  //it will create a snapshot of search component and match it with a prev snapshot
  test("has valid snapshot", () => {
    const component = renderer.create(<Search />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Button", () => {
  it("renders success", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Button>Button component</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has valid snapshot", () => {
    const component = renderer.create(<Button />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Table", () => {
  const props = {
    list: [
      { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
      { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" },
    ],
    sortKey: "TITLE",
  };

  it("renders success", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  //enzyme test
  it("shows items in list", () => {
    const element = shallow(<Table {...props} />);
    expect(element.find(".table-row").length).toBe(2);
  });
});
