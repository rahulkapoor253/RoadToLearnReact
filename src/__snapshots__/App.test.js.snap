// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has valid snapshot 1`] = `
<button
  className=""
  type="button"
/>
`;

exports[`Search has valid snapshot 1`] = `
<form>
  <input
    maxLength="15"
    type="text"
  />
  <button
    className="btn-search"
    type="submit"
  >
    Submit
  </button>
</form>
`;
